clippy_lints/src/unnecessary_blocking_ops.rs

```diff
@@ -0,0 +1,182 @@
+use rustc_data_structures::fx::FxHashMap;
+use rustc_errors::{Applicability, Diagnostic};
+use rustc_hir::def_id::DefId;
+use rustc_hir::{
+    Body, BodyId, Closure, ClosureKind, CoroutineDesugaring, CoroutineKind, Expr, ExprKind, ImplItem, ImplItemKind,
+    Item, ItemKind, Node, TraitItem, TraitItemKind,
+};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for async function or async closure with blocking operations that
+    /// could be replaced with their async counterpart.
+    ///
+    /// ### Why is this bad?
+    /// Blocking a thread prevents tasks being swapped, causing other tasks to stop running
+    /// until the thread is no longer blocked, which might not be as expected in an async context.
+    ///
+    /// ### Known problems
+    /// Not all blocking operations can be detected, as for now, this lint only detects a small
+    /// set of functions in standard library by default. And some of the suggestions might need
+    /// additional features to work properly.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::time::Duration;
+    /// pub async fn foo() {
+    ///     std::thread::sleep(Duration::from_secs(5));
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```ignore
+    /// use std::time::Duration;
+    /// pub async fn foo() {
+    ///     tokio::time::sleep(Duration::from_secs(5));
+    /// }
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub UNNECESSARY_BLOCKING_OPS,
+    pedantic,
+    "blocking operations in an async context"
+}
+
+pub(crate) struct UnnecessaryBlockingOps {
+    blocking_ops: Vec<DisallowedPath>,
+    /// Map of resolved funtion def_id with suggestion string after checking crate
+    id_with_suggs: FxHashMap<DefId, Option<String>>,
+    /// Tracking whether a body is async after entering it.
+    body_asyncness: Vec<bool>,
+}
+
+impl UnnecessaryBlockingOps {
+    pub(crate) fn new(blocking_ops: Vec<DisallowedPath>) -> Self {
+        Self {
+            blocking_ops,
+            id_with_suggs: FxHashMap::default(),
+            body_asyncness: vec![],
+        }
+    }
+}
+
+impl_lint_pass!(UnnecessaryBlockingOps => [UNNECESSARY_BLOCKING_OPS]);
+
+static HARD_CODED_BLOCKING_OP_PATHS: &[&str] = &[
+    "std::thread::sleep",
+    // Filesystem functions
+    "std::fs::try_exists",
+    "std::fs::canonicalize",
+    "std::fs::copy",
+    "std::fs::create_dir",
+    "std::fs::create_dir_all",
+    "std::fs::hard_link",
+    "std::fs::metadata",
+    "std::fs::read",
+    "std::fs::read_dir",
+    "std::fs::read_link",
+    "std::fs::read_to_string",
+    "std::fs::remove_dir",
+    "std::fs::remove_dir_all",
+    "std::fs::remove_file",
+    "std::fs::rename",
+    "std::fs::set_permissions",
+    "std::fs::symlink_metadata",
+    "std::fs::write",
+    // IO functions
+    "std::io::copy",
+    "std::io::read_to_string",
```

Comment:
These are only blocking if the underlying reader or writer are blocking. For example, copying between `Vec`s or reading from a `Vec` into a string isn't blocking in practice.

Reply:
> ...only blocking if the underlying reader or writer are blocking.

That's interesting, can you further explain this? 

Initially, I list those methods purely base on the observation from the `tokio` crate, since there are `AsyncRead` and `AsyncWrite` in the crate, I assume the `Read` and `Write` traits from `std::io` are blocking :smile: , but now that you mention it, I wonder if copying between slices would actually affect anything.